walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,150 @@
 import java.util.Collection;
+import java.util.Iterator; 
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requestAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<String> allAttendees = new ArrayList<>();
+    allAttendees.addAll(requestAttendees);
+    allAttendees.addAll(optionalAttendees);
+    long requestDuration = request.getDuration();
+    int numRequestSlots = (int) requestDuration / 30;
+    int numEvents = events.size();
+
+    boolean[] mandatoryTimeSlots = new boolean[48];
+    boolean[] allTimeSlots = new boolean[48];
+
+    // if the duration is longer than a day there are no options
+    if (requestDuration >= TimeRange.WHOLE_DAY.duration() + 1)
+      return Arrays.asList();
+    // if there are no attendees, the time range is the whole day
+    if (events.size() == 0 || allAttendees.size() == 0) 
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    
+    // sets up timeSlots array; anywhere left true has all the requestAttendees available
+    for (Event event : events) { 
+      mandatoryTimeSlots = convertToTimeSlots(requestAttendees, events);
+      allTimeSlots = convertToTimeSlots(allAttendees, events);    
+    }
+
+    ArrayList<Integer> possibleSlots = convertToSlots(mandatoryTimeSlots, numRequestSlots);
+    ArrayList<Integer> allSlots = convertToSlots(allTimeSlots, numRequestSlots);
+    
+    ArrayList<TimeRange> possibleTimes = convertToTimes(possibleSlots, numRequestSlots, numEvents);
+    ArrayList<TimeRange> allTimes = convertToTimes(allSlots, numRequestSlots, numEvents);
+
+    if (allTimes.isEmpty() && !possibleTimes.contains(TimeRange.WHOLE_DAY))
+        return possibleTimes;
+    return allTimes;
+  }
+  
+  public static boolean[] convertToTimeSlots (Collection<String> requestAttendees, Collection<Event> events) {
+    boolean[] timeSlots = new boolean[48];
+    Arrays.fill(timeSlots, Boolean.TRUE);
+    int numEvents = events.size();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      TimeRange when = event.getWhen();
+      int start = (when.start()) / 30;
+      int eventSlots = when.duration() / 30;
+
+      Iterator<String> eventAi = eventAttendees.iterator(); 
+      while(eventAi.hasNext()) {
+        numEvents--;
+        String eventAttendee = eventAi.next();
+
+        Iterator<String> requestAi = requestAttendees.iterator(); 
+        while (requestAi.hasNext()) {
+          String requestAttendee = requestAi.next();
+          if (eventAttendee.equals(requestAttendee)) {
+            for (int i = 0; i < eventSlots; i++) 
+              timeSlots[start + i] = false;
+            break;
+          }
+        }
+        if(numEvents == 0)
+            break;
```

Comment:
Can you explain why you need to have this check here? At the beginning, `numEvents` is set to events.size(), but then you seem to be decreasing it for every eventAttendee so I'm not sure what this if statement is checking for.

Reply:
it's to increase the efficiency of the algorithm 
if the next number isn't +1 from the last number, then that means there's an event in between and it decreases the numEvents count
if there are no more events to look for then there is no need to keep checking and I can just add the rest of the free time to possibleTimes
this doesn't always work, for example, in cases where events are right next to each other or overlapping, but it works in enough cases to reduce the time elapsed by almost half (from ~0.1 to ~0.06 seconds)

Reply:
actually nvm I ran mvn test a couple more times and the time elapsed is all over the place
so it's just to make me feel better about the algorithm efficiency I guess lol